File: clients/python/provider/superposition_provider/data_sources/file_data_source.py
"""File-based implementation of SuperpositionDataSource."""

import os
from pathlib import Path
from typing import Dict, List, Optional, Any

from superposition_bindings.superposition_client import ffi_parse_toml_config

from ..interfaces import SuperpositionDataSource
from ..types import ConfigData, ExperimentData


class FileDataSource(SuperpositionDataSource):
    """File-based data source for Superposition configuration.

    Reads configuration from a local file (TOML or JSON format).
    Uses native bindings for TOML parsing.
    """

    def __init__(self, file_path: str):
        """Initialize FileDataSource.

        Args:
            file_path: Path to the configuration file
        """
        self.file_path = Path(file_path)

    async def fetch_config(self) -> ConfigData:
        """Fetch the latest configuration from the data source."""
        try:
            content = self.file_path.read_text(encoding="utf-8")
            ext = self.file_path.suffix.lower()

            if ext == ".toml":
                # Use native TOML parser from bindings
                parsed = ffi_parse_toml_config(content)
                return ConfigData(
                    default_configs=parsed.default_configs or {},
                    contexts=parsed.contexts or [],
                    overrides=parsed.overrides or {},
                    dimensions=parsed.dimensions or {},
                )
            elif ext == ".json":
                # Use JSON parser for JSON files
                import json

                data = json.loads(content)
                return ConfigData(
                    default_configs=data.get("default_configs", {}),
                    contexts=data.get("contexts", []),
                    overrides=data.get("overrides", {}),
                    dimensions=data.get("dimensions", {}),
                )
            else:
                # Try TOML first, then JSON
                try:
                    parsed = ffi_parse_toml_config(content)
                    return ConfigData(
                        default_configs=parsed.default_configs or {},
                        contexts=parsed.contexts or [],
                        overrides=parsed.overrides or {},
                        dimensions=parsed.dimensions or {},
                    )
                except Exception:
                    # Try JSON if TOML fails
                    import json

                    data = json.loads(content)
                    return ConfigData(
                        default_configs=data.get("default_configs", {}),
                        contexts=data.get("contexts", []),
                        overrides=data.get("overrides", {}),
                        dimensions=data.get("dimensions", {}),
                    )
        except Exception as e:
            raise RuntimeError(
                f"Failed to read or parse config file at {self.file_path}: {e}"
            )

    async def fetch_filtered_config(
        self,
        context: Optional[Dict[str, Any]] = None,
        prefix_filter: Optional[List[str]] = None,
    ) -> ConfigData:
        """Fetch configuration with context/prefix filters."""
        config = await self.fetch_config()

        filtered_contexts = config.contexts
        if context:
            filtered_contexts = [
                ctx for ctx in config.contexts if self._matches_context(ctx, context)
            ]

        filtered_configs = config.default_configs
        if prefix_filter:
            filtered_configs = self._filter_by_prefix(
                config.default_configs, prefix_filter
            )

        return ConfigData(
            default_configs=filtered_configs,
            contexts=filtered_contexts,
            overrides=config.overrides,
            dimensions=config.dimensions,
        )

    async def fetch_active_experiments(self) -> Optional[ExperimentData]:
        """Fetch all active experiment data.

        File data source doesn't support experiments.
        """
        return None

    async def fetch_candidate_active_experiments(
        self,
        context: Optional[Dict[str, Any]] = None,
        prefix_filter: Optional[List[str]] = None,
    ) -> Optional[ExperimentData]:
        """Fetch active experiments filtered with partial context matching.

        File data source doesn't support experiments.
        """
        return None

    async def fetch_matching_active_experiments(
        self,
        context: Optional[Dict[str, Any]] = None,
        prefix_filter: Optional[List[str]] = None,
    ) -> Optional[ExperimentData]:
        """Fetch active experiments filtered with exact context matching.

        File data source doesn't support experiments.
        """
        return None

    def supports_experiments(self) -> bool:
        """Check if this data source supports experiments."""
        return False

    async def close(self) -> None:
        """Close and cleanup resources.

        No cleanup needed for file data source.
        """
        pass

    def _matches_context(
        self,
        context: Dict[str, Any],
        filter_dict: Dict[str, Any],
    ) -> bool:
        """Check if context matches filter criteria."""
        for key, value in filter_dict.items():
            if context.get(key) != value:
                return False
        return True

    def _filter_by_prefix(
        self,
        configs: Dict[str, Any],
        prefixes: List[str],
    ) -> Dict[str, Any]:
        """Filter config keys by prefixes."""
        filtered: Dict[str, Any] = {}
        for key, value in configs.items():
            if any(key.startswith(prefix) for prefix in prefixes):
                filtered[key] = value
        return filtered

File: clients/javascript/open-feature-provider/data-sources/file-data-source.ts
import { readFile } from "fs/promises";
import { existsSync } from "fs";
import path from "path";
import {
    SuperpositionDataSource,
    ConfigData,
    ExperimentData,
} from "../types";
import { NativeResolver } from "superposition-bindings";

/**
 * File-based implementation of SuperpositionDataSource
 * Reads configuration from a local file (JSON or TOML format)
 */
export class FileDataSource implements SuperpositionDataSource {
    private filePath: string;
    private nativeResolver: NativeResolver;

    constructor(filePath: string, nativeResolver?: NativeResolver) {
        this.filePath = filePath;
        this.nativeResolver = nativeResolver || new NativeResolver();
    }

    /**
     * Fetch the latest configuration from the data source
     */
    async fetchConfig(): Promise<ConfigData> {
        try {
            const content = await readFile(this.filePath, "utf-8");
            const ext = path.extname(this.filePath).toLowerCase();

            if (ext === ".toml") {
                // Use native TOML parser from superposition-bindings
                const parsed = this.nativeResolver.parseTomlConfig(content);
                return {
                    default_configs: parsed.default_configs || {},
                    contexts: parsed.contexts || [],
                    overrides: parsed.overrides || {},
                    dimensions: parsed.dimensions || {},
                    fetched_at: new Date(),
                };
            } else if (ext === ".json") {
                // Use JSON parser for JSON files
                const data = JSON.parse(content);
                return {
                    default_configs: data.default_configs || {},
                    contexts: data.contexts || [],
                    overrides: data.overrides || {},
                    dimensions: data.dimensions || {},
                    fetched_at: new Date(),
                };
            } else {
                // Try JSON first, then TOML
                try {
                    const data = JSON.parse(content);
                    return {
                        default_configs: data.default_configs || {},
                        contexts: data.contexts || [],
                        overrides: data.overrides || {},
                        dimensions: data.dimensions || {},
                        fetched_at: new Date(),
                    };
                } catch {
                    // Try TOML if JSON fails
                    const parsed = this.nativeResolver.parseTomlConfig(content);
                    return {
                        default_configs: parsed.default_configs || {},
                        contexts: parsed.contexts || [],
                        overrides: parsed.overrides || {},
                        dimensions: parsed.dimensions || {},
                        fetched_at: new Date(),
                    };
                }
            }
        } catch (error) {
            throw new Error(
                `Failed to read or parse config file at ${this.filePath}: ${error}`
            );
        }
    }

    /**
     * Fetch configuration with context/prefix filters
     */
    async fetchFilteredConfig(
        context?: Record<string, any>,
        prefixFilter?: string[]
    ): Promise<ConfigData> {
        const config = await this.fetchConfig();

        let filteredContexts = config.contexts;
        if (context && Object.keys(context).length > 0) {
            filteredContexts = config.contexts.filter((ctx) =>
                this.matchesContext(ctx, context)
            );
        }

        let filteredConfigs = config.default_configs;
        if (prefixFilter && prefixFilter.length > 0) {
            filteredConfigs = this.filterByPrefix(
                config.default_configs,
                prefixFilter
            );
        }

        return {
            default_configs: filteredConfigs,
            contexts: filteredContexts,
            overrides: config.overrides,
            dimensions: config.dimensions,
            fetched_at: config.fetched_at,
        };
    }

    /**
     * Fetch all active experiment data
     * File data source doesn't support experiments
     */
    async fetchActiveExperiments(): Promise<ExperimentData | null> {
        return null;
    }

    /**
     * Fetch active experiments filtered with partial context matching
     * File data source doesn't support experiments
     */
    async fetchCandidateActiveExperiments(
        _context?: Record<string, any>,
        _prefixFilter?: string[]
    ): Promise<ExperimentData | null> {
        return null;
    }

    /**
     * Fetch active experiments filtered with exact context matching
     * File data source doesn't support experiments
     */
    async fetchMatchingActiveExperiments(
        _context?: Record<string, any>,
        _prefixFilter?: string[]
    ): Promise<ExperimentData | null> {
        return null;
    }

    /**
     * Check if this data source supports experiments
     */
    supportsExperiments(): boolean {
        return false;
    }

    /**
     * Close and cleanup resources
     */
    async close(): Promise<void> {
        // No cleanup needed for file data source
    }

    /**
     * Check if context matches filter criteria
     */
    private matchesContext(
        context: Record<string, any>,
        filter: Record<string, any>
    ): boolean {
        for (const [key, value] of Object.entries(filter)) {
            if (context[key] !== value) {
                return false;
            }
        }
        return true;
    }

    /**
     * Filter config keys by prefixes
     */
    private filterByPrefix(
        configs: Record<string, any>,
        prefixes: string[]
    ): Record<string, any> {
        const filtered: Record<string, any> = {};

        for (const [key, value] of Object.entries(configs)) {
            if (prefixes.some((prefix) => key.startsWith(prefix))) {
                filtered[key] = value;
            }
        }

        return filtered;
    }
}
